### PR TITLE
feat(CX-3339): New phone input component

### PR DESCRIPTION
**Pull request by MrSltun**
THIS PR TO BE MERGED TO `mrsltun/CX-3330/new-input-component`!

This PR resolves [CX-3339]

### Description
This PR adds a new component for Palette called PhoneInput which requires a list of country code list in a specific structure typed with TypeScript
The new component displays the list and allows filtering by searching for the country name or the country code number

### Screenshots
| Status | Screenshot |
| --- | --- |
| Empty | image |
| With a placeholder | image |
| Disabled | image |
| Filled | image |
| Hover | image |
| Focused | image |
| Required | image |
| With the list open | image |
| With the list open and filtered | image |
| With errors | image |

### Demo

| Jounrey | Video |
| -- | -- |
| ⌨️ & 🖱️ | <video src="https://user-images.githubusercontent.com/20655703/233412661-97e58cc1-2632-4229-ac42-f739c2cd3774.mov" width="670" /> | 
| ⌨️ ONLY | <video src="https://user-images.githubusercontent.com/20655703/233412825-04ad00c7-db69-4047-a50f-12a24c0fa5c4.mov" /> |

[CX-3339]: https://artsyproduct.atlassian.net/browse/CX-3339?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/palette/src/elements/PhoneInput/PhoneInput.tsx [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            reset()
+
+            set({ cursor: 0, interactive: true })
+          }
+          return
+
+        default:
+          searchInputRef.current?.focus()
+      }
+    }
```
[the remark this answers is not in the document]
Comment (dzucconi) [reply packages/palette/src/elements/PhoneInput/PhoneInput.tsx:231]: Or maybe a better question is why that component couldn't be used in this case?

[the remark this answers is not in the document]
Comment (dzucconi) [reply packages/palette/src/elements/PhoneInput/PhoneInput.tsx:231]: (Oh, I see it's like a persistent filterable list, vs a dropdown)

File: packages/palette/src/elements/Input/Input.tsx [hunk 2]
```diff
@@ -143,7 +143,7 @@ const StyledInput = styled.input<StyledInputProps>`
   padding: 0 ${themeGet("space.1")};
   appearance: none;
   line-height: 1;
-  border: 1px solid ${themeGet("colors.black15")};
```
Comment (olerichter00) [inline packages/palette/src/elements/Input/Input.tsx:146]: **question:** Is this change intentional?

File: packages/palette/src/elements/PhoneInput/PhoneInput.tsx [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        >
+          <SelectContainer
+            ref={countryPickerRef as any}
+            disabled={disabled}
+            onClick={() => {
+              if (!disabled && !isDropdownVisible) {
+                setDropdownVisible(true)
+              } else {
+                setDropdownVisible(false)
+              }
```
Comment (olerichter00) [inline packages/palette/src/elements/PhoneInput/PhoneInput.tsx:258]: **suggestion:**

```suggestion
              setDropdownVisible(!disabled && !isDropdownVisible)
```

File: packages/palette/src/elements/PhoneInput/PhoneInput.tsx [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+            autoComplete="tel-national"
+            name={inputName}
+            maxLength={25}
+            placeholder={inputProps.placeholder}
+            onChange={(e) => {
+              const inputValue = e.target.value
+              const countryCodeValue = selectedOption.value
+              inputProps.onChange?.(e)
+
+              onInputChange?.({ inputValue, countryCodeValue })
```
Comment (olerichter00) [inline packages/palette/src/elements/PhoneInput/PhoneInput.tsx:279]: **suggestion:**

I am not sure if this makes a difference, but updating the internal state first might make the UI update faster (in the case `inputProps.onChange` takes a long time).

 ```suggestion
               onInputChange?.({ inputValue, countryCodeValue })

               inputProps.onChange?.(e)
```

**Review by olerichter00 — COMMENTED** [review]
This looks great 🌟 

Just added a few small comments.

File: packages/palette/src/elements/PhoneInput/PhoneInput.tsx [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+              })}
+            </SelectOptions>
+          </SelectDropdown>
+        )}
+
+        {required && !(error && typeof error === "string") && (
+          <Text mt={0.5} ml={1} variant="xs" color="black60" textAlign="left">
+            *Required
+          </Text>
+        )}
```
Comment (dariakoko) [inline packages/palette/src/elements/PhoneInput/PhoneInput.tsx:325]: Is there a *Required component we can reuse?

**Review by dariakoko — APPROVED** [review]
LGTM! Thanks Sultan

File: packages/palette/src/elements/Input/Input.tsx (hunk 2 above)
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:146]: Yes actually, since we don't have `black15` in our design system anymore :)

File: packages/palette/src/elements/PhoneInput/PhoneInput.tsx (hunk 5 above)
Comment (MrSltun) [reply packages/palette/src/elements/PhoneInput/PhoneInput.tsx:325]: No we don't but that's a good idea :) I'll add it now!

File: packages/palette/src/shared/RequiredField.tsx [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import React from "react"
+import { Text, TextProps } from "../elements"
+
+export const RequiredField: React.FC<TextProps> = (props) => {
+  return (
+    <Text variant="xs" color="black60" textAlign="left" {...props}>
+      *Required
+    </Text>
```
Comment (dzucconi) [inline packages/palette/src/shared/RequiredField.tsx:8]: This is not how we denote required fields in the design system, and I don't really buy the reasoning for this being a component.

Comment (dzucconi) [reply packages/palette/src/shared/RequiredField.tsx:8]: Oh, duh! Sorry, I totally forgot the context of this PR!

Comment (dzucconi) [reply packages/palette/src/shared/RequiredField.tsx:8]: (Of course, feel free to leave it if that's universally how we denote required state now)

**Comment by MrSltun** [comment]
I'll merge this PR to `mrsltun/CX-3330/new-input-component` within the next 2 hours to update the [review app](https://new-input.artsy.net/) on force to test it! Please let me know if you have any objections :)

**Comment by artsyit** [comment]
:rocket: PR was released in [`@artsy/palette-charts@36.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette-charts%4036.0.0), [`@artsy/palette@37.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette%4037.0.0) :rocket:
